blog/2023-10-22-websockets-app.md

```diff
@@ -0,0 +1,484 @@
+    ws.on('error', console.error);
+    
+    ws.on('message', function message(data) {
+      onMessage(ws.id, data.toString("utf8"));
+    });
+
+    ws.on('close', function () {
+      onDisconnect(ws.id);
+    });
+
+    onConnect(ws.id);
+  });
+  
+  return {
+    close: () => {
+      console.log("closing server...");
+      wss.close();
+    },
+    url: () => `ws://127.0.0.1:${port}`
+  }
+};
+```
+
+Note: make sure to install the `ws` NPM library first. 
+
+In this code we export a function which initiates a web server. The function accepts our inflight handlers and calls them when the relevant
+events occurs. Note that we set a connection ID on the socket itself so we can interact with those connections later.
+Let's add the missing javascript part - sending a message to a connection:
+
+```js
+export const sendMessage = (connectionId, message) => {
+  let wss = global.wss;
+  if (!wss) {
+    return;
+  }
+
+  wss.clients.forEach((ws) => {
+    if (ws.id !== connectionId) {
+      return;
+    }
+
+    ws.send(message)
+  });
+}
+```
+
+Now, we can write our local implementation and connect our Winglang code to the JS exports using the `extern` keyword:
+
+```js
+pub class WebSocketApi_sim impl IWebsocketsApi {
+  extern "./websockets-api-local.js" static inflight startWebSocketApi(
+    connectFn: inflight (str): void,
+    disconnectFn: inflight (str): void,
+    onmessageFn: inflight (str, str): void
+  ): StartWebSocketApiResult;
+  extern "./websockets-api-local.js" static inflight sendMessage(
+    connectionId: str,
+    message: str,
+  ): inflight(): void;
+}
+```
+
+To start the server we can use the `cloud.Service` resource which will run once the simulator starts:
+
+```js
+pub class WebSocketApi_sim impl IWebsocketsApi {
+  var connectFn: inflight (str): void;
+  var disconnectFn: inflight (str): void;
+  var onmessageFn: inflight (str, str): void;
+  _url: sim.State;
+
+  init() {
+    this.connectFn = inflight () => {};
+    this.disconnectFn = inflight () => {};
+    this.onmessageFn = inflight () => {};
+    this._url = new sim.State();
+    new cloud.Service(inflight () => {
+      let res = WebSocketApi_sim.startWebSocketApi(this.connectFn, this.disconnectFn, this.onmessageFn);
+      this._url.set("service_url", res.url());
+      return () => {
+        res.close();
+      };
+    });
+  }
+
+  pub onConnect(fn: inflight (str): void) {
+    this.connectFn = fn;
+  }
+
+  pub onDisconnect(fn: inflight (str): void) {
+    this.disconnectFn = fn;
+  }
+
+  pub onMessage(fn: inflight (str, str): void) {
+    this.onmessageFn = fn;
+  }
+
+  extern "./websockets-api-local.js" static inflight startWebSocketApi(
+    connectFn: inflight (str): void,
+    disconnectFn: inflight (str): void,
+    onmessageFn: inflight (str, str): void
+  ): StartWebSocketApiResult;
+  extern "./websockets-api-local.js" static inflight sendMessage(
+    connectionId: str,
+    message: str,
+  ): inflight(): void;
+}
```

Comment:
I would recommend consolidating all extern calls to `ws` under a single module called `ws.w` with a `Util` class. This has the undocumented side effect that you can call the static methods without the class name. It also makes it cleaner from a code organization perspective.

Something like this:

`ws.w`

```js
pub class Util {
  extern "./ws.js" pub static inflight startServer(
    connectFn: inflight (str): void,
    disconnectFn: inflight (str): void,
    onmessageFn: inflight (str, str): void
  ): StartWebSocketApiResult;

  extern "./ws.js" pub static inflight sendMessage(
    connectionId: str,
    message: str,
  ): inflight(): void;
}
```
 
Then:

`main.w`:

```js
bring "./ws.w" as ws;

ws.startServer(...)
```

Yes, it works...